### PR TITLE
WIP: Error handling

### PR DESCRIPTION
**Pull request by hseeberger**
Work in progress, don't merge!

### CODE REVIEW

File: streamee-demo/src/main/scala/io/moia/streamee/demo/TextShuffler.scala [hunk 1]
```diff
@@ -62,13 +69,21 @@ object TextShuffler {
         Await.result(wordShufflerSinkRef().map(_.sink), wordShufflerAskTimeout) // Hopefully we can get rid of blocking soon: https://github.com/akka/akka/issues/25934
       }
 
-    Process[ShuffleText, TextShuffled]
-      .via(delayRequest(delay))
-      .via(keepSplitShuffle(wordShufflerSink, wordShufflerProcessorTimeout))
-      .via(concat)
+    Process[ShuffleText, Either[Error, TextShuffled]]
+      .via(validateRequest)
+      .viaEither(delayProcessing(delay))
```
[the remark this answers is not in the document]
Comment (hseeberger) [reply streamee-demo/src/main/scala/io/moia/streamee/demo/TextShuffler.scala]: Much much better! Will change.

File: streamee/src/main/scala/io/moia/streamee/package.scala [hunk 2]
```diff
@@ -195,10 +204,47 @@ package object streamee {
...
+
+          val bcast     = builder.add(Broadcast[(Either[E, Out], CtxOut)](2, eagerCancel = true))
+          val merge     = builder.add(Merge[(Either[E, Out2], CtxOut)](2, eagerComplete = true))
+          val leftOnly  = FlowWithContext[Either[E, Out], CtxOut].collect { case Left(e) => Left(e) }
+          val rightOnly = FlowWithContext[Either[E, Out], CtxOut].collect { case Right(out) => out }
+          val mapRight  = FlowWithContext[Out2, CtxOut].map(Right(_))
+
+          // format: OFF
+        flowWithContext ~> bcast ~> leftOnly             ~>             merge
+                           bcast ~> rightOnly ~> viaFlow ~> mapRight ~> merge
```
[the remark this answers is not in the document]
Comment (hseeberger) [reply streamee/src/main/scala/io/moia/streamee/package.scala]: Yeah, there is no more "let's merge two things togehter". Just two independent routes for success and failure.

File: streamee/src/main/scala/io/moia/streamee/package.scala [hunk 3]
```diff
@@ -195,10 +204,69 @@ package object streamee {
...
+        // format: ON
+
+          FlowShape(flowWithContext.in, merge.out)
+        })
+      FlowWithContext.fromTuples(flow)
+    }
+
+    def rightFlatVia[Out2](
```
Comment (hseeberger) [inline streamee/src/main/scala/io/moia/streamee/package.scala]: @aquamatthias I have not found a way to overload `rightVia`, hence I had to add a second name for this method.

Two questions:
- Do you see a way to overload a single method name?
- If not, what about different names, e.g. `mapVia` (the current `rightVia`) and `flatMapVia` (this method)?

BTW, see `TextShuffler` above for using both methods together.

File: streamee/src/main/scala/io/moia/streamee/package.scala [hunk 4]
```diff
@@ -195,10 +204,69 @@ package object streamee {
...
   }
 
+  /**
+    * Extension methods for `FlowWithContext` with paired output context.
+    */
+  final implicit class EitherFlowWithContextOps[In, CtxIn, Out, CtxOut, Mat, E](
+      val flowWithContext: FlowWithContext[In, CtxIn, Either[E, Out], CtxOut, Mat]
+  ) extends AnyVal {
+
+    def mapEither[Out2](f: Out => Out2): FlowWithContext[In, CtxIn, Either[E, Out2], CtxOut, Any] =
```
Comment (aquamatthias) [inline streamee/src/main/scala/io/moia/streamee/package.scala]: Consider to name this method `mapRight`.

File: streamee/src/main/scala/io/moia/streamee/package.scala (hunk 3 above)
Comment (aquamatthias) [reply streamee/src/main/scala/io/moia/streamee/package.scala]: I think it is fine to not overload the name and be precise about the expected flow.
In terms of name: 
- `via` is the akka streams name to connect flows
- map and flatMap operate on the values of the flow
- either is right biased, we could omit the term right. 

At first glance I stumbled upon the combination of `via` and `map`. 
But I have to admit that it "speaks to me" and is very clear. 
So I like your proposal of `mapVia` and `flatMapVia` - let's use it.

Comment (hseeberger) [reply streamee/src/main/scala/io/moia/streamee/package.scala]: Thanks!

**Pull request merged**